doc/server_settings.rst
######################################
Server Settings
######################################

******************************
Passing settings to the client
******************************

If you start a Loom server without specifying any settings, Loom uses appropriate defaults for a local Loom server.

::
   
   loom server start

Non-default settings are assigned to a new server in one of three ways: using the ``--settings-file``/``-s`` flag to specify one or more config files:

::

   loom server start --settings-file mysettings.conf

or individually, using the ``--extra-settings``/``-e`` flag:

::

   loom server start --extra-settings LOOM_SERVER_NAME=MyServerName

or through environment variables that use the ``LOOM_`` prefix:

::

   export LOOM_SERVER_NAME=MyServerName
   loom server start

Multiple ``--settings-file`` flags are allowed, with the last file given highest priority for any conflicts. Multiple ``--extra-settings`` flags are also allowed. If different sources or settings are used together, the order of precedence, from highest to lowest, is:

1. environment variable
2. ``--extra-settings``
3. ``--settings-file``

********************
Settings file format
********************

Loom settings files are formatted like ini files, but without section headers. They are a flat file with a list of key-value pairs, separated by ":" or "=" and optional spaces. Comments are denoted with "#", and blank lines are allowed. For example:

*demosettings.conf:*

::

   LOOM_SERVER_NAME: loom-demo
   
   # Google cloud settings
   LOOM_MODE: gcloud
   LOOM_GCLOUD_SERVER_INSTANCE_TYPE: n1-standard-4
   LOOM_GCLOUD_WORKER_BOOT_DISK_SIZE_GB: 50

*********
Resources
*********

Some settings give the name of a file resource used by Loom. For example, LOOM_SSL_CERT_KEY_FILE and LOOM_SSL_CERT_FILE refer to the SSL key and certificate, and LOOM_GCE_PEM_FILE refers to the service account key file for Google Cloud Platform.

Rather than hard-code the path to these files, you must place them in one directory and make the setting indicate the file name or relative path in that directory. The directory is specified with the ``--resources`` flag on "loom server start".

*ssl-settings.conf* file contents:

::

   LOOM_SSL_CERT_KEY_FILE: loom.key
   LOOM_SSL_CERT_FILE: loom.crt

*resources/* directory contents:

::

   resources/loom.key
   resources/loom.crt

Use this command to start the server with the necessary files:

::

   loom server start --resource-dir ./resources --settings-file ssl-settings.conf 

On the client machine, Loom will cache a copy of these files in ~/.loom (or the value of the environment variable LOOM_SETTINGS_HOME), and it will create an additional copy on the server if it is remote, so the original copy does not need to be retained for the server to work.

*******************************************
Required settings to launch on Google Cloud
*******************************************

To launch a server on Google Cloud Platform, at a minimum you will need these required settings:

LOOM_MODE: gcloud
LOOM_GCE_PEM_FILE: your-gcp-key.json
LOOM_GCE_EMAIL: service-account-id@your-gcp-project.iam.gserviceaccount.com
LOOM_GCE_PROJECT: your-gcp-project
LOOM_GOOGLE_STORAGE_BUCKET: your-gcp-bucket

*****************
Updating settings
*****************

Not all settings can be changed safely. Changing LOOM_MYSQL_DATABASE, for example, will leave Loom unable to connect to the database.

However, sometimes it is necessary to modify settings, and many of them, such as LOOM_DEBUG, can be safely changed. You may want to test changes in a staging server before working with production data.

On the client machine:

1. Edit setting(s) in ~/.loom/server-settings.conf
2. Edit any resource files in ~/.loom/resources/
3. Wait for any runs in progress to complete.
4. loom server stop
5. loom server start

This will overwrite settings and resource files on the server with any changes you have made.

*****************
Modes
*****************

The LOOM_MODE setting is used to toggle between deployment modes. Currently loom has two modes: "local" and "gcloud". The default mode is "local".

*****************
Custom playbooks
*****************

The primary difference between modes is that they correspond to different sets of playbooks. You can find these in the Loom source code under "loomengine/client/playbooks/". Loom requires at least these five playbooks to be defined for any mode:

* {mode}_cleanup_task_attempt.yml
* {mode}_delete_server.yml
* {mode}_run_task_attempt.yml
* {mode}_start_server.yml
* {mode}_stop_server.yml

So for example, in the playbooks directory you will see a "gcloud_stop_server.yml" and a "local_stop_server.yml".

Loom allows you to use a custom set of playbooks to control how Loom is deployed. To do this, first create a copy of the "loomengine/client/playbooks" directory. Use the "local_*.yml" or "gcloud_*.yml" playbooks as a starting point. You may wish to change the prefix, but make sure that when you launch a new server, the LOOM_MODE setting matches the playbook prefix that you choose.

To pass the custom playbooks directory to loom when starting a new server, use the ``--playbooks`` flag:

::

   loom server start --my-custom-settings.conf --playbook-dir ./my-custom-playbooks

Loom settings are passed to the playbooks as environment variables. You are welcome to use your own settings for custom playbooks, but you may have to disable settings validation with "LOOM_SKIP_SETTINGS_VALIDATION=true".

*******************
Index of settings
*******************

----------------

================ ================
*default*        
*valid values*   
*notes*          
================ ================


Settings for all modes
**********************

LOOM_SERVER_NAME
----------------

================ ================
*default*        loom-server
*valid values*   String that begins with alpha, ends with alphanumeric, and contains only alphanumeric or -. Max length of 63.
================ ================

LOOM_SERVER_NAME determines how several components of the Loom server named. For example, the Docker container hosting the Loom server web application is named {{LOOM_SERVER_NAME}}-master, and the instance hosting the server in gcloud mode is named {{LOOM_SERVER_NAME}}.

LOOM_MODE
----------------

================ ================
*default*        local
*valid values*   local|gcloud|{custom}
================ ================

LOOM_MODE selects between different sets of playbooks. It also changes some default settings and the rules for settings validation. Supported modes are "local" and "gcloud". You may also develop custom playbooks that are compatible with another mode.

LOOM_DEBUG
----------------

================ ================
*default*        false
*valid values*   true|false
================ ================

When true, it activates several debugging tools and verbose server errors. Primarily for development use.

LOOM_LOG_LEVEL
----------------

================ ================
*default*        INFO
*valid values*   CRITICAL|ERROR|WARNING|INFO|DEBUG
================ ================

LOOM_DOCKER_IMAGE
-----------------

================ ================
*default*        loomengine/loom:{version}
================ ================

Docker image for server, worker, and scheduler

LOOM_DEFAULT_DOCKER_REGISTRY
----------------

================ ================
*default*        none
================ ================

LOOM_DEFAULT_DOCKER_REGISTRY applies to the LOOM_DOCKER_IMAGE and "docker_image" values in templates. Anywhere that a repo is given with no specific registry, LOOM_DEFAULT_DOCKER_REGISTRY will be assumed.

LOOM_STORAGE_TYPE
----------------

================ ================
*default*        local
*valid values*   local|google_storage
================ ================

Sets the type of persistent file storage. Usually google_storage would only be used with gcloud mode, but Loom does not impose this restriction. This may be useful for testing or for a custom deployment mode.

LOOM_STORAGE_ROOT
-----------------

================== ================
*default (local)*  ~/loomdata
*default (gcloud)* /loomdata
*valid values*     absolute file path
================== ================

LOOM_GOOGLE_STORAGE_BUCKET
--------------------------

================ ================
*default*        None. Setting is required if LOOM_STORAGE_TYPE==google_storage
*valid values*   Valid Google Storage bucket name.
================ ================

Loom will attempt to create the bucket it if it does not exist.

LOOM_ANSIBLE_INVENTORY
----------------------

================== ================
*default (local)*  localhost,
*default (gcloud)* gce_inventory_wrapper.py
*valid values*     Comma-delimited list of hosts, or executable filename
================== ================

Accepts either a comma-separated list of host inventory (e.g. "localhost," -- the comma is required) or a dynamic inventory executable. The executable must be in the playbooks directory.
 
LOOM_ANSIBLE_HOST_KEY_CHECKING
------------------------------

================ ================
*default*        false
*valid values*   true|false
================ ================

Leaving LOOM_ANSIBLE_HOST_KEY_CHECKING as false will ignore warnings about invalid host keys. These errors are common on Google Cloud Platform where IP addresses are frequently reused, causing conflicts with known_hosts.

LOOM_HTTP_PORT
--------------

================ ================
*default*        80
*valid values*   1–65535
================ ================

LOOM_HTTPS_PORT
---------------

================ ================
*default*        443
*valid values*   1–65535
================ ================

LOOM_HTTP_PORT_ENABLED
----------------------

================ ================
*default*        true
*valid values*   true|false
================ ================

LOOM_HTTPS_PORT_ENABLED
-----------------------

================ ================
*default*        false
*valid values*   true|false
================ ================

LOOM_HTTP_REDIRECT_TO_HTTPS
---------------------------

================ ================
*default*        false
*valid values*   true|false
================ ================

If true, NGINX will redirect requests on LOOM_HTTP_PORT to LOOM_HTTPS_PORT.

LOOM_SSL_CERT_KEY_FILE
----------------------

================ ================
*default*        {{LOOM_SERVER_NAME}}+'-ssl-cert-key.pem'
================ ================

LOOM_SSL_CERT_FILE
------------------

================ ================
*default*        {{LOOM_SERVER_NAME}}+'-ssl-cert.pem'
================ ================

LOOM_SSL_CERT_CREATE_NEW
------------------------

================ ================
*default*        false
*valid values*   true|false
================ ================

If true, Loom will create a self-signed certificate and key. If LOOM_SSL_CERT_CREATE_NEW==false and LOOM_HTTPS_PORT_ENABLED==true, user must provide certificate and key in the resources directory and set LOOM_SSL_CERT_KEY_FILE and LOOM_SSL_CERT_FILE to the correct filenames.

LOOM_SSL_CERT_C
----------------

================ ================
*default*        US
================ ================

Used in subject field if self-signed SSL certificate if LOOM_SSL_CERT_CREATE_NEW==true.

LOOM_SSL_CERT_ST
----------------

================ ================
*default*        California
================ ================

Used in subject field if self-signed SSL certificate if LOOM_SSL_CERT_CREATE_NEW==true.

LOOM_SSL_CERT_L
----------------

================ ================
*default*        Palo Alto
================ ================

Used in subject field if self-signed SSL certificate if LOOM_SSL_CERT_CREATE_NEW==true.

LOOM_SSL_CERT_O
----------------

================ ================
*default*        Stanford University
================ ================

Used in subject field if self-signed SSL certificate if LOOM_SSL_CERT_CREATE_NEW==true.

LOOM_SSL_CERT_CN
----------------

================ ================
*default*        {{ansible_hostname}}
================ ================

Used in subject field if self-signed SSL certificate if LOOM_SSL_CERT_CREATE_NEW==true.

LOOM_SERVER_ALLOWED_HOSTS
-------------------------

================ ================
*default*        [*]
================ ================

List of hosts from which Loom will accept a connection. Corresponds to the django ALLOWED_HOSTS setting.

LOOM_SERVER_CORS_ORIGIN_ALLOW_ALL
---------------------------------

================ ================
*default*        false
================ ================

Whitelist all hosts for cross-origin resource sharing. Corresponds to the django CORS_ORIGIN_ALLOW_ALL setting.

LOOM_SERVER_CORS_ORIGIN_WHITELIST
---------------------------------

================ ================
*default*        []
================ ================

Hosts to be whitelisted for cross-origin resource sharing. Corresponds to the django CORS_ORIGIN_WHITELIST setting.

LOOM_TASKRUNNER_HEARTBEAT_INTERVAL_SECONDS
------------------------------------------

================ ================
*default*        60
================ ================

Frequency of heatbeats sent by TaskAttempt monitor process to Loom server.

LOOM_TASKRUNNER_HEARTBEAT_TIMEOUT_SECONDS
-----------------------------------------

================ ================
*default*        300
================ ================

Kill any TaskAttempt that has not sent a heartbeat in this time.

LOOM_MAXIMUM_TASK_RETRIES
-------------------------

================ ================
*default*        2
================ ================

Maximum number of TaskAttempt retries.

LOOM_PRESERVE_ON_FAILURE
------------------------

================ ================
*default*        false
*valid values*   true|false
================ ================

Do not clean up instance or containers for any failed TaskAttempts. May be useful for debugging.

LOOM_PRESERVE_ALL
-----------------

================ ================
*default*        false
*valid values*   true|false
================ ================

Do not clean up instance or containers for any TaskAttempts. May be useful for debugging.

LOOM_SERVER_GUNICORN_WORKERS_COUNT
----------------------------------

================ ================
*default*        10
================ ================

LOOM_WORKER_CELERY_CONCURRENCY
------------------------------

================ ================
*default*        30
================ ================

LOOM_MYSQL_CREATE_DOCKER_CONTAINER
----------------------------------

================ ================
*default*        true
================ ================

Create a new Docker container to host the Loom database instead of connecting to an external database.

LOOM_MYSQL_HOST
----------------

================ ================
*default*        {{mysql_container_name}} if LOOM_MYSQL_CREATE_DOCKER_CONTAINER==true; otherwise no default.
================ ================

MySQL server connection settings.

LOOM_MYSQL_PORT
----------------

================ ================
*default*        3306
================ ================

MySQL server connection settings.

LOOM_MYSQL_USER
----------------

================ ================
*default*        loom
================ ================

MySQL server connection settings.

LOOM_MYSQL_PASSWORD
-------------------

================ ================
*default*        loompass
================ ================

MySQL server connection settings.

LOOM_MYSQL_DATABASE
-------------------

================ ================
*default*        loomdb
================ ================

MySQL server connection settings.

LOOM_MYSQL_IMAGE
----------------

================ ================
*default*        mysql:5.7.17
================ ================

Docker image used to create MySQL container if LOOM_MYSQL_CREATE_DOCKER_CONTAINER==true.

LOOM_MYSQL_RANDOM_ROOT_PASSWORD
-------------------------------

================ ================
*default*        true
================ ================

Create a random root password when initializing database if LOOM_MYSQL_CREATE_DOCKER_CONTAINER==true.

LOOM_MYSQL_SSL_CA_CERT_FILE
---------------------------

================ ================
*default*        none
================ ================

If needed, certificate files for MySQL database connection should be provided through the resources directory.

LOOM_MYSQL_SSL_CLIENT_CERT_FILE
-------------------------------

================ ================
*default*        none
================ ================


doc/releasenotes.rst
#############
Release Notes
#############

Latest
=============

0.6.0
------------
* Installable from PyPi with "pip install loomengine"

Previous
=============

0.5.3
------------
* Retry tasks on failure or lost heartbeat
* Paginated index views in browser
* Chunked index view queries by commandline client
* Migrated kibana logs to same port as Loom under /logs
* Critical bugfix for corrupted data trees
* Simplified output filenames and directory structure
* Periodic check for tasks that failed to clean up
* Client warns, not errors, for duplicate file or template imports
* Add setting LOOM_FORCE_DB_MIGRATIONS_ON_START

0.5.2
------------
* "loom server start" now starts a local server using default settings
* Required settings to launch on gcloud were reduced to a minimal set
* Server settings validation
* Removed deprecated commands "loom show", "loom import", and "loom export"
* Removed deprecated fixed_inputs in templates
* Deprecation warning for "loom server start --admin-files-dir", renamed flag as "--resource-dir"
* Server settings can be changed without destroying server, using "loom server start/stop"
* Docker image versions for Loom server components can be upgraded without destroying server, using "loom server start/stop"

0.5.1
-------------
* Enhanced validation of templates and data objects

0.5.0
-------------
* Tags and labels for files, templates, and runs
* Changed client commands to follow 'loom {noun} {verb}' pattern

0.4.1
-------------
* Notification for completed runs by email or posting JSON to URL
* Documentation for Loom templates
* Templates can be referenced by hash
* Added retries to file import/export, docker pull, other external services
* Added ``--original-copy`` option to "loom file import"
* Added LOOM_DEFAULT_DOCKER_REGISTRY setting

0.4.0
-------------
* Parallel workflows
* Deprecated fixed inputs, replaced with optional and overridable "data" field on standard inputs
* User-defined run names, using the optional ``--name`` flag with "loom run"
* Updated examples, including two parallel examples "word_scoring" and "word_combinations"
* Saving of templates is no longer asynchronous, so any errors are raised immediately with "loom import template"
* Outputs can now use "glob" source in addition to "filename" and "stream"

0.3.8
-------------
* Run overview shows nested runs, tasks, and task-attempts

0.3.7
-------------
* Retries for upload/download from Google Storage

0.3.6
-------------
* Runs have "waiting" status until they start
* Runs are no longer changed to "killed" if they already completed
* Input/output detail routes on runs

0.3.5
-------------
* Critical bugfix for 0.3.4

0.3.4
-------------
* Pre-configure Kibana
* Disable X-Pack features in Kibana and Elasticsearch
* Handle several sporadic failures from gcloud services
* Handle gcloud gucket to bucket file copies longer than 30s
* Prune docker data volumes

0.3.3
-------------
* Critical bugfix for 0.3.2 that prevented use on Google Cloud

0.3.2
-------------
* Fluentd for logging, with kibana+elasticsearch for log viewing
* Nested templates by reference
* API documentation with swagger
* Reduced lag time in running tasks

0.3.1
-------------
* Allow substitution in template output filenames
* Added LOOM_PRESERVE_ON_FAILURE and LOOM_PRESERVE_ALL flags for debugging
* Several bugfixes

0.3.0
-------------
* User-configurable playbooks
* Non-reverse-compatible simplifications to API
* Reduced server response times
* Dockerized deployment on local and google cloud
* Optional dockerized MySQL server
* Retry tasks if process stops responding

0.2.1
-------------
* Use release-specific DOCKER_TAG in default settings

0.2.0
-------------
* Loom can create a server locally or on Google Cloud Platform
* Accepts workflow templates in JSON or YAML format
* Web portal provides a brower interface for viewing templates, files, and runs
* Loom client for managing runs from the terminal


server/loomengine_server/core/settings.py
# Quick-start development settings - unsuitable for production
# See https://docs.djangoproject.com/en/1.7/howto/deployment/checklist/

import datetime
import json
import logging
import os
import random
import socket
import sys
import tempfile
import warnings

def to_boolean(value):
    if value in [None, '', False]:
        return False
    if value == True:
        return True
    if str(value).lower() == 'false':
        return False
    if str(value).lower() == 'true':
        return True
    raise Exception("Invalid value %s. Expected True or False")

def to_float(value):
    if value is None:
        return None
    if value == '':
        return None
    return float(value)

def to_int(value):
    if value is None:
        return None
    if value == '':
        return None
    return int(value)

def to_list(value):
    if value is None:
        return []
    value = value.strip(' "\'')
    list_str = value.lstrip('[').rstrip(']')
    if list_str == '':
        return []
    list = list_str.split(',')
    return [item.strip(' "\'') for item in list]

SETTINGS_DIR = os.path.dirname(__file__)
BASE_DIR = (os.path.join(SETTINGS_DIR, '..'))
sys.path.append(BASE_DIR)
PORTAL_ROOT = os.path.join(BASE_DIR, 'portal')

# Security settings
DEBUG = to_boolean(os.getenv('LOOM_DEBUG'))
SECRET_KEY = os.getenv(
    'LOOM_SERVER_SECRET_KEY',
    ''.join([random.SystemRandom()\
             .choice('abcdefghijklmnopqrstuvwxyz0123456789!@#$%^&*(-_=+)')
             for i in range(50)]))
CORS_ORIGIN_ALLOW_ALL = to_boolean(
    os.getenv('LOOM_SERVER_CORS_ORIGIN_ALLOW_ALL', 'False'))
CORS_ORIGIN_WHITELIST = to_list(os.getenv('LOOM_SERVER_CORS_ORIGIN_WHITELIST', '[]'))

ALLOWED_HOSTS = to_list(os.getenv('LOOM_SERVER_ALLOWED_HOSTS', '[*]'))

LOG_LEVEL = os.getenv('LOG_LEVEL', 'WARNING').upper()

STORAGE_TYPE = os.getenv('LOOM_STORAGE_TYPE', 'local').lower()

STATIC_ROOT = os.getenv('LOOM_SERVER_STATIC_ROOT', '/var/www/loom/static')

SERVER_NAME = os.getenv('LOOM_SERVER_NAME', 'loom') # used in attempt container names
SERVER_URL_FOR_WORKER = os.getenv('SERVER_URL_FOR_WORKER', 'http://127.0.0.1:8000')
SERVER_URL_FOR_CLIENT = os.getenv('SERVER_URL_FOR_CLIENT', 'http://127.0.0.1:8000')
STORAGE_ROOT = os.path.expanduser(os.getenv('LOOM_STORAGE_ROOT', '~/loom-data'))
FILE_ROOT_FOR_WORKER = os.path.expanduser(
    os.getenv('FILE_ROOT_FOR_WORKER', STORAGE_ROOT))

TASKRUNNER_HEARTBEAT_INTERVAL_SECONDS = int(os.getenv('LOOM_TASKRUNNER_HEARTBEAT_INTERVAL_SECONDS', '60'))
TASKRUNNER_HEARTBEAT_TIMEOUT_SECONDS = int(os.getenv('LOOM_TASKRUNNER_HEARTBEAT_TIMEOUT_SECONDS', TASKRUNNER_HEARTBEAT_INTERVAL_SECONDS*2.5))
SYSTEM_CHECK_INTERVAL_MINUTES = int(os.getenv('LOOM_SYSTEM_CHECK_INTERVAL_MINUTES', '15'))
PRESERVE_ON_FAILURE = to_boolean(os.getenv('LOOM_PRESERVE_ON_FAILURE', 'False'))
PRESERVE_ALL = to_boolean(os.getenv('LOOM_PRESERVE_ALL', 'False'))
MAXIMUM_RETRIES_FOR_ANALYSIS_FAILURE = int(os.getenv(
    'LOOM_MAXIMUM_TASK_RETRIES_FOR_ANALYSIS_FAILURE', '1'))
MAXIMUM_RETRIES_FOR_SYSTEM_FAILURE = int(os.getenv(
    'LOOM_MAXIMUM_TASK_RETRIES_FOR_SYSTEM_FAILURE', '10'))

DEFAULT_DOCKER_REGISTRY = os.getenv('LOOM_DEFAULT_DOCKER_REGISTRY', '')

# GCP settings
GCE_EMAIL = os.getenv('GCE_EMAIL')
GCE_PROJECT = os.getenv('GCE_PROJECT', '')
GCE_PEM_FILE_PATH = os.getenv('GCE_PEM_FILE_PATH')
GOOGLE_STORAGE_BUCKET = os.getenv('LOOM_GOOGLE_STORAGE_BUCKET', '')

WORKER_BOOT_DISK_TYPE = os.getenv('WORKER_BOOT_DISK_TYPE')
WORKER_BOOT_DISK_SIZE = os.getenv('WORKER_BOOT_DISK_SIZE')
WORKER_CUSTOM_SUBNET = os.getenv('WORKER_CUSTOM_SUBNET', '')
WORKER_LOCATION = os.getenv('WORKER_LOCATION')
WORKER_NETWORK = os.getenv('WORKER_NETWORK')
WORKER_SCRATCH_DISK_MOUNT_POINT = os.getenv('WORKER_SCRATCH_DISK_MOUNT_POINT')
WORKER_SCRATCH_DISK_TYPE = os.getenv('WORKER_SCRATCH_DISK_TYPE')
WORKER_SCRATCH_DISK_SIZE = os.getenv('WORKER_SCRATCH_DISK_SIZE')
WORKER_SKIP_INSTALLS = os.getenv('WORKER_SKIP_INSTALLS')
WORKER_TAGS = os.getenv('WORKER_TAGS')
WORKER_USES_SERVER_INTERNAL_IP = os.getenv('WORKER_USES_SERVER_INTERNAL_IP')
WORKER_VM_IMAGE = os.getenv('WORKER_VM_IMAGE')

SETTINGS_HOME = os.getenv('LOOM_SETTINGS_HOME', os.path.expanduser('~/.loom'))
PLAYBOOK_PATH = os.path.join(SETTINGS_HOME, os.getenv('LOOM_PLAYBOOK_DIR', 'playbooks'))
RUN_TASK_ATTEMPT_PLAYBOOK = os.getenv('LOOM_RUN_TASK_ATTEMPT_PLAYBOOK')
CLEANUP_TASK_ATTEMPT_PLAYBOOK = os.getenv('LOOM_CLEANUP_TASK_ATTEMPT_PLAYBOOK')

# Database settings
# Any defaults must match defaults in playbook
MYSQL_HOST = os.getenv('LOOM_MYSQL_HOST')
MYSQL_USER = os.getenv('LOOM_MYSQL_USER', 'loom')
MYSQL_PASSWORD = os.getenv('LOOM_MYSQL_PASSWORD', 'loompass')
MYSQL_DATABASE = os.getenv('LOOM_MYSQL_DATABASE', 'loomdb')
MYSQL_PORT = int(os.getenv('LOOM_MYSQL_PORT', 3306))
MYSQL_SSL_CA_CERT_PATH = os.getenv('LOOM_MYSQL_SSL_CA_CERT_PATH')
MYSQL_SSL_CLIENT_CERT_PATH = os.getenv('LOOM_MYSQL_SSL_CLIENT_CERT_PATH')
MYSQL_SSL_CLIENT_KEY_PATH = os.getenv('LOOM_MYSQL_SSL_CLIENT_KEY_PATH')

# Email settings
EMAIL_BACKEND = 'django.core.mail.backends.smtp.EmailBackend'
EMAIL_HOST = os.getenv('LOOM_EMAIL_HOST', None)
EMAIL_PORT = to_int(os.getenv('LOOM_EMAIL_PORT', 587))
EMAIL_HOST_USER = os.getenv('LOOM_EMAIL_HOST_USER', None)
EMAIL_HOST_PASSWORD = os.getenv('LOOM_EMAIL_HOST_PASSWORD', None)
EMAIL_USE_TLS = to_boolean(os.getenv('LOOM_EMAIL_USE_TLS', True))
EMAIL_USE_SSL = to_boolean(os.getenv('LOOM_EMAIL_USE_SSL', True))
EMAIL_TIMEOUT = to_float(os.getenv('LOOM_EMAIL_TIMEOUT', 0.0))
EMAIL_SSL_KEYFILE = os.getenv('LOOM_EMAIL_SSL_KEYFILE', None)
EMAIL_SSL_CERTFILE = os.getenv('LOOM_EMAIL_SSL_CERTFILE', None)
DEFAULT_FROM_EMAIL = os.getenv('LOOM_DEFAULT_FROM_EMAIL', EMAIL_HOST_USER)
NOTIFICATION_ADDRESSES = to_list(os.getenv('LOOM_NOTIFICATION_ADDRESSES', '[]'))
NOTIFICATION_HTTPS_VERIFY_CERTIFICATE = to_boolean(os.getenv('LOOM_NOTIFICATION_HTTPS_VERIFY_CERTIFICATE', True))

# Message broker settings
LOOM_RABBITMQ_PASSWORD = os.getenv('LOOM_RABBITMQ_PASSWORD', 'guest')
LOOM_RABBITMQ_USER = os.getenv('LOOM_RABBITMQ_USER', 'guest')
LOOM_RABBITMQ_VHOST = os.getenv('LOOM_RABBITMQ_VHOST', '/')
LOOM_RABBITMQ_HOST = os.getenv('LOOM_RABBITMQ_HOST', 'rabbitmq')
LOOM_RABBITMQ_PORT = os.getenv('LOOM_RABBITMQ_PORT', '5672')

def _get_ansible_inventory():
    ansible_inventory = os.getenv('LOOM_ANSIBLE_INVENTORY', 'localhost,')
    if ',' not in ansible_inventory:
	ansible_inventory = os.path.join(
            PLAYBOOK_PATH,
            os.getenv('LOOM_ANSIBLE_INVENTORY'))
    return ansible_inventory

ANSIBLE_INVENTORY = _get_ansible_inventory()
LOOM_SSH_PRIVATE_KEY_PATH = os.getenv('LOOM_SSH_PRIVATE_KEY_PATH')

KEEP_DUPLICATE_FILES = True
FORCE_RERUN = True

# For testing only
TEST_DISABLE_ASYNC_DELAY = to_boolean(os.getenv('TEST_DISABLE_ASYNC_DELAY', False))
TEST_NO_CREATE_TASK = to_boolean(os.getenv('TEST_NO_CREATE_TASK', False))
TEST_NO_RUN_TASK_ATTEMPT = to_boolean(os.getenv('TEST_NO_RUN_TASK_ATTEMPT', False))
TEST_NO_POSTPROCESS = to_boolean(os.getenv('TEST_NO_POSTPROCESS', False))
TEST_NO_PUSH_INPUTS_ON_RUN_CREATION = to_boolean(
    os.getenv('TEST_NO_PUSH_INPUTS_ON_RUN_CREATION', False))

# Fixed settings
SECURE_PROXY_SSL_HEADER = ('HTTP_X_FORWARDED_PROTO', 'https')
LANGUAGE_CODE = 'en-us'
TIME_ZONE = 'UTC'
USE_TZ = True
CELERY_ALWAYS_EAGER = True
APPEND_SLASH = True
ROOT_URLCONF = 'loomengine_server.core.urls'

# Celery
CELERY_RESULT_BACKEND = 'django-cache'
CELERY_BROKER_URL = 'amqp://%s:%s@%s:%s/%s' \
                    % (LOOM_RABBITMQ_USER, LOOM_RABBITMQ_PASSWORD,
                       LOOM_RABBITMQ_HOST, LOOM_RABBITMQ_PORT,
                       LOOM_RABBITMQ_VHOST)
CELERY_BROKER_POOL_LIMIT = 50
CELERYD_TASK_SOFT_TIME_LIMIT = 60

INSTALLED_APPS = [
    'django.contrib.auth',
    'django.contrib.contenttypes',
    'django_extensions',
    'django.contrib.sessions',
    'django.contrib.messages',
    'django.contrib.staticfiles',
    'corsheaders',
    'rest_framework',
    'rest_framework_swagger',
    'django_celery_results',
    'api',
    'mptt',
]

MIDDLEWARE_CLASSES = [
    'django.contrib.sessions.middleware.SessionMiddleware',
    'corsheaders.middleware.CorsMiddleware',
    'django.middleware.common.CommonMiddleware',
    'django.middleware.csrf.CsrfViewMiddleware',
#    'django.contrib.auth.middleware.AuthenticationMiddleware',
#    'django.contrib.auth.middleware.SessionAuthenticationMiddleware',
    'django.contrib.messages.middleware.MessageMiddleware',
    'django.middleware.clickjacking.XFrameOptionsMiddleware',
]

REST_FRAMEWORK = {
    # Use Django's standard `django.contrib.auth` permissions,
    # or allow read-only access for unauthenticated users.
    'DEFAULT_PERMISSION_CLASSES': (
        'rest_framework.permissions.AllowAny',
    ),
    'DEFAULT_RENDERER_CLASSES': (
        'rest_framework.renderers.JSONRenderer',
        'rest_framework.renderers.BrowsableAPIRenderer',
    ),
    'DEFAULT_PAGINATION_CLASS': 'rest_framework.pagination.LimitOffsetPagination',
}

TEMPLATES = [
    {
        'BACKEND': 'django.template.backends.django.DjangoTemplates',
        'DIRS': [os.path.join(BASE_DIR, "templates")],
        'APP_DIRS': True,
        'OPTIONS': {
            'context_processors': [
                'django.template.context_processors.debug',
                'django.template.context_processors.request',
                'django.contrib.auth.context_processors.auth',
                'django.contrib.messages.context_processors.messages',
            ],
        },
    },
]

WSGI_APPLICATION = 'loomengine_server.core.wsgi.application'

def _get_sqlite_databases():
    return {
        'default': {
            'ENGINE': 'django.db.backends.sqlite3',
            'NAME': os.path.join(BASE_DIR, 'loomdb.sqlite3'),
        }
    }

def _get_mysql_databases():
    if not MYSQL_USER:
        raise Exception(
            "LOOM_MYSQL_USER is a required setting if LOOM_MYSQL_HOST is set")
    if not MYSQL_DATABASE:
        raise Exception(
            "LOOM_MYSQL_DATABASE is a required setting if LOOM_MYSQL_HOST is set")

    DATABASES = {
        'default': {
            'ENGINE': 'django.db.backends.mysql',
            'HOST': MYSQL_HOST,
            'NAME': MYSQL_DATABASE,
            'USER': MYSQL_USER,
            'PORT': MYSQL_PORT,
        }
    }

    if MYSQL_PASSWORD:
        DATABASES['default'].update({
            'PASSWORD': MYSQL_PASSWORD
        })
    if MYSQL_SSL_CA_CERT_PATH \
       or MYSQL_SSL_CLIENT_CERT_PATH \
       or MYSQL_SSL_CLIENT_KEY_PATH:
        if not (MYSQL_SSL_CA_CERT_PATH \
                and MYSQL_SSL_CLIENT_CERT_PATH \
                and MYSQL_SSL_CLIENT_KEY_PATH):
            raise Exception(
                'One or more required values missing: '\
                'LOOM_MYSQL_SSL_CA_CERT_PATH="%s", '\
                'LOOM_MYSQL_SSL_CLIENT_CERT_PATH="%s", '\
                'LOOM_MYSQL_SSL_CLIENT_KEY_PATH="%s"' % (
                    MYSQL_SSL_CA_CERT_PATH,
                    MYSQL_SSL_CLIENT_CERT_PATH,
                    MYSQL_SSL_CLIENT_KEY_PATH))
        else:
            DATABASES['default'].update({
                'OPTIONS': {
                    'ssl': {
                        'ca': MYSQL_SSL_CA_CERT_PATH,
                        'cert': MYSQL_SSL_CLIENT_CERT_PATH,
                        'key': MYSQL_SSL_CLIENT_KEY_PATH
                    }
                }
            })
    return DATABASES

# Database
if MYSQL_HOST:
    DATABASES = _get_mysql_databases()
else:
    DATABASES = _get_sqlite_databases()

# Logging
if len(sys.argv) > 1 and sys.argv[1] == 'test':
    DISABLE_LOGGING = True
else:
    DISABLE_LOGGING = False

if DISABLE_LOGGING:
    LOGGING = {}
else:
    LOGGING = {
        'version': 1,
        'disable_existing_loggers': False,
        'formatters': {
            'default': {
                'format': '%(levelname)s [%(asctime)s] %(message)s'
            },
        },
        'handlers': {
            'console': {
                'class': 'logging.StreamHandler',
                'formatter': 'default',
            }
        },
        'loggers': {
            'django': {
                'handlers': ['console'],
                'level': LOG_LEVEL,
            },
            'loomengine': {
                'handlers': ['console'],
                'level': LOG_LEVEL,
            },
            'api': {
                'handlers': ['console'],
                'level': LOG_LEVEL,
            },
        },
    }

STATIC_URL = '/%s/' % os.path.basename(STATIC_ROOT)
STATICFILES_DIRS = [
    os.path.join(BASE_DIR, "static"),
]

CSRF_COOKIE_SECURE = True
SESSION_COOKIE_SECURE = True

INTERNAL_IPS = [
    "127.0.0.1",
]

if DEBUG or (len(sys.argv) > 1 and sys.argv[1] == 'collectstatic'):
    INSTALLED_APPS.append('debug_toolbar')
    MIDDLEWARE_CLASSES.append('debug_toolbar.middleware.DebugToolbarMiddleware')

    def custom_show_toolbar(request):
        return True

    DEBUG_TOOLBAR_CONFIG = {
        "INTERCEPT_REDIRECTS": False,
        'MEDIA_URL': '/__debug__/m/',
        'SHOW_TOOLBAR_CALLBACK': custom_show_toolbar,
    }

CACHES = {
    'default': {
        'BACKEND': 'django.core.cache.backends.locmem.LocMemCache',
        'LOCATION': '232871b2',
        'TIMEOUT': 0,
    }
}
